File: src/shared/regions.py
from __future__ import annotations

import math
from dataclasses import dataclass

from PIL import Image

@dataclass(frozen=True)
class ImageRegion:
    left: int
    top: int
    width: int
    height: int

    @property
    def right(self) -> int:
        return self.left + self.width

    @property
    def bottom(self) -> int:
        return self.top + self.height

    def as_box(self) -> tuple[int, int, int, int]:
        return (self.left, self.top, self.right, self.bottom)

@dataclass(frozen=True)
class UniformTileGrid:
    total_width: int
    total_height: int
    tile_width: int
    tile_height: int
    columns: int
    rows: int

    @property
    def padded_width(self) -> int:
        return self.tile_width * self.columns

    @property
    def padded_height(self) -> int:
        return self.tile_height * self.rows

    def iter_regions(self):
        for row in range(self.rows):
            for col in range(self.columns):
                left = col * self.tile_width
                top = row * self.tile_height
                yield row, col, ImageRegion(
                    left=left,
                    top=top,
                    width=min(self.tile_width, self.padded_width - left),
                    height=min(self.tile_height, self.padded_height - top),
                )

def build_uniform_tile_grid(
    total_width: int,
    total_height: int,
    *,
    max_tile_width: int,
    max_tile_height: int | None = None,
    min_tiles_per_axis: int = 1,
) -> UniformTileGrid:
    safe_total_width = max(1, int(total_width))
    safe_total_height = max(1, int(total_height))
    safe_max_tile_width = max(1, int(max_tile_width))
    safe_max_tile_height = max(
        1,
        int(max_tile_height if max_tile_height is not None else max_tile_width),
    )
    safe_min_tiles = max(1, int(min_tiles_per_axis))

    columns = max(
        safe_min_tiles,
        int(math.ceil(safe_total_width / float(safe_max_tile_width))),
    )
    rows = max(
        safe_min_tiles,
        int(math.ceil(safe_total_height / float(safe_max_tile_height))),
    )

    tile_width = int(math.ceil(safe_total_width / float(columns)))
    tile_height = int(math.ceil(safe_total_height / float(rows)))

    return UniformTileGrid(
        total_width=safe_total_width,
        total_height=safe_total_height,
        tile_width=tile_width,
        tile_height=tile_height,
        columns=columns,
        rows=rows,
    )

def build_square_tile_grid(
    total_width: int,
    total_height: int,
    *,
    max_tile_extent: int,
    min_tiles_per_axis: int = 1,
) -> UniformTileGrid:
    safe_total_width = max(1, int(total_width))
    safe_total_height = max(1, int(total_height))
    safe_max_extent = max(1, int(max_tile_extent))
    safe_min_tiles = max(1, int(min_tiles_per_axis))

    divisions = max(
        safe_min_tiles,
        int(
            math.ceil(
                max(safe_total_width, safe_total_height) / float(safe_max_extent)
            )
        ),
    )
    tile_width = int(math.ceil(safe_total_width / float(divisions)))
    tile_height = int(math.ceil(safe_total_height / float(divisions)))

    return UniformTileGrid(
        total_width=safe_total_width,
        total_height=safe_total_height,
        tile_width=tile_width,
        tile_height=tile_height,
        columns=divisions,
        rows=divisions,
    )

def pad_image_to_size(
    image: Image.Image,
    width: int,
    height: int,
    *,
    fill: tuple[int, int, int, int] = (0, 0, 0, 0),
) -> Image.Image:
    target_width = max(1, int(width))
    target_height = max(1, int(height))
    if image.width == target_width and image.height == target_height:
        return image

    result = Image.new("RGBA", (target_width, target_height), fill)
    result.paste(image, (0, 0))
    return result

def compute_centered_box(
    *,
    width: int,
    height: int,
    center_x: float,
    center_y: float,
    box_width: float,
    box_height: float | None = None,
) -> tuple[float, float, float, float]:
    safe_width = max(1.0, float(width))
    safe_height = max(1.0, float(height))
    target_box_width = max(1.0, min(float(box_width), safe_width))
    target_box_height = max(
        1.0,
        min(
            float(box_height if box_height is not None else box_width),
            safe_height,
        ),
    )

    half_w = target_box_width / 2.0
    half_h = target_box_height / 2.0

    left = min(max(0.0, float(center_x) - half_w), max(0.0, safe_width - target_box_width))
    top = min(max(0.0, float(center_y) - half_h), max(0.0, safe_height - target_box_height))
    right = left + target_box_width
    bottom = top + target_box_height

    return (left, top, right, bottom)

File: cpp/app/shared/image_processing/regions.cpp
#include "shared/image_processing/regions.h"

#include <algorithm>
#include <cmath>

namespace imgsli::app::shared::image_processing {

std::vector<UniformTileGrid::Tile> UniformTileGrid::tiles() const {
  std::vector<Tile> out;
  out.reserve(static_cast<std::size_t>(rows * columns));
  const int padW = paddedWidth();
  const int padH = paddedHeight();
  for (int row = 0; row < rows; ++row) {
    for (int col = 0; col < columns; ++col) {
      const int left = col * tileWidth;
      const int top = row * tileHeight;
      out.push_back(Tile{row, col,
                         ImageRegion{left, top,
                                     std::min(tileWidth, padW - left),
                                     std::min(tileHeight, padH - top)}});
    }
  }
  return out;
}

namespace {

int ceilDiv(int a, int b) {
  return static_cast<int>(std::ceil(static_cast<double>(a) / static_cast<double>(b)));
}

}  // namespace

UniformTileGrid buildUniformTileGrid(int totalWidth, int totalHeight,
                                     int maxTileWidth, int maxTileHeight,
                                     int minTilesPerAxis) {
  const int sw = std::max(1, totalWidth);
  const int sh = std::max(1, totalHeight);
  const int mtw = std::max(1, maxTileWidth);
  const int mth = std::max(1, maxTileHeight > 0 ? maxTileHeight : maxTileWidth);
  const int sm = std::max(1, minTilesPerAxis);

  const int cols = std::max(sm, ceilDiv(sw, mtw));
  const int rows = std::max(sm, ceilDiv(sh, mth));
  return UniformTileGrid{sw, sh, ceilDiv(sw, cols), ceilDiv(sh, rows), cols, rows};
}

UniformTileGrid buildSquareTileGrid(int totalWidth, int totalHeight,
                                    int maxTileExtent, int minTilesPerAxis) {
  const int sw = std::max(1, totalWidth);
  const int sh = std::max(1, totalHeight);
  const int me = std::max(1, maxTileExtent);
  const int sm = std::max(1, minTilesPerAxis);
  const int divisions = std::max(sm, ceilDiv(std::max(sw, sh), me));
  return UniformTileGrid{sw, sh, ceilDiv(sw, divisions), ceilDiv(sh, divisions),
                         divisions, divisions};
}

QRectF computeCenteredBox(int width, int height, double centerX, double centerY,
                          double boxWidth, double boxHeight) {
  const double sw = std::max(1.0, static_cast<double>(width));
  const double sh = std::max(1.0, static_cast<double>(height));
  const double bw = std::clamp(boxWidth, 1.0, sw);
  const double effectiveBoxH = boxHeight > 0.0 ? boxHeight : boxWidth;
  const double bh = std::clamp(effectiveBoxH, 1.0, sh);
  const double halfW = bw / 2.0;
  const double halfH = bh / 2.0;
  const double left =
      std::min(std::max(0.0, centerX - halfW), std::max(0.0, sw - bw));
  const double top =
      std::min(std::max(0.0, centerY - halfH), std::max(0.0, sh - bh));
  return QRectF(left, top, bw, bh);
}

}  // namespace imgsli::app::shared::image_processing
